File: Frontend/AktuelUrunBulucu/src/api/searchApi.js
const API_URL = import.meta.env.VITE_API_URL

export async function getUserIp() {
  const res = await fetch('https://api.ipify.org?format=json')
  const data = await res.json()
  return data.ip
}

export async function searchProducts(query, ip) {
  const res = await fetch(`${API_URL}/api/search?query=${encodeURIComponent(query)}&ip=${encodeURIComponent(ip)}`)
  if (!res.ok) {
    let title = 'Bir sorun çıktı'
    let detail = 'Bilinmeyen hata'
    try {
      const body = await res.json()
      if (body.title) title = body.title
      if (body.detail) detail = body.detail
    } catch {
      // JSON parse edilemezse default mesajları kullan
    }
    const err = new Error(detail)
    err.title = title
    throw err
  }
  return res.json()
}

export async function saveUserLocation(ip, latitude, longitude) {
  await fetch(`${API_URL}/api/location`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ ip, latitude, longitude })
  })
}

export async function subscribeNotification(ip, email, product) {
  const res = await fetch(`${API_URL}/api/notify`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ ip, email, product })
  })
  if (!res.ok) throw new Error('Bildirim kaydedilemedi')
}

File: Mobile/lib/api/search_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../utils/constants.dart';
import '../models/product_result.dart';

Future<String> getUserIp() async {
  final res = await http.get(Uri.parse(ipifyUrl));
  return jsonDecode(res.body)['ip'] as String;
}

Future<({bool found, List<ProductResult> products})> searchProducts(
    String query, String ip) async {
  final uri = Uri.parse('$apiBaseUrl/api/search')
      .replace(queryParameters: {'query': query, 'ip': ip});
  final res = await http.get(uri);
  if (res.statusCode != 200) throw Exception('API hatası');
  final json = jsonDecode(res.body);
  final found = json['found'] as bool;
  final products = found
      ? (json['matchedProducts'] as List)
          .map((e) => ProductResult.fromJson(e))
          .toList()
      : <ProductResult>[];
  return (found: found, products: products);
}

Future<void> saveUserLocation(String ip, double lat, double lon) async {
  await http.post(
    Uri.parse('$apiBaseUrl/api/location'),
    headers: {'Content-Type': 'application/json'},
    body: jsonEncode({'ip': ip, 'latitude': lat, 'longitude': lon}),
  );
}

Future<void> subscribeNotification(
    String ip, String email, String product) async {
  final res = await http.post(
    Uri.parse('$apiBaseUrl/api/notify'),
    headers: {'Content-Type': 'application/json'},
    body: jsonEncode({'ip': ip, 'email': email, 'product': product}),
  );
  if (res.statusCode != 200) throw Exception('Bildirim kaydedilemedi');
}
